### PR TITLE
Add support for client acknowledgements and QoS settings 

### LINKED ISSUE #14842: Follow security best practices in RabbitMQ connector
**Description:**
Follow security best practices like specifying sensitive parameters and tainted return values.

**Suggested Assignees:**
Arshika

### PR DESCRIPTION
**Pull request by aashikam**
## Purpose
Add support for manual and auto acknowledgement modes and allow setting global and consumer service specific QoS settings. 

Fixes #14842

## Samples
**Consumer with client acknowledgement**
```ballerina
import ballerina/log;
import ballerina/rabbitmq;

rabbitmq:Connection connection = new({ host: "localhost", port: 5672 });
listener rabbitmq:ChannelListener channelListener = new(connection);

@rabbitmq:ServiceConfig {
    queueConfig: {
        queueName: "MyQueue"
    },
    ackMode: rabbitmq:CLIENT_ACK
}
service testClientAcknowledgement on channelListener {
    resource function onMessage(rabbitmq:Message message) {
        var msg = message.getTextContent();
        if (msg is string) {
            log:printInfo("The message received: " + msg);
        } else {
            log:printError("Error occurred while retrieving the message content.");
        }
        // Positively acknowledges a single message.
        var result = message->basicAck();
    }
}
```

**Consumer with QoS settings**
```ballerina
import ballerina/log;
import ballerina/rabbitmq;

rabbitmq:Connection connection = new({ host: "localhost", port: 5672 });

listener rabbitmq:ChannelListener channelListener = new(connection);

@rabbitmq:ServiceConfig {
    queueConfig: {
        queueName: "MyQueue"
    },
    ackMode: rabbitmq:CLIENT_ACK,
    prefetchCount: 10
}
service testClientQos on channelListener {
    resource function onMessage(rabbitmq:Message message) {
        var msg = message.getTextContent();
        if (msg is string) {
            log:printInfo("The message received: " + msg);
        } else {
            log:printError("Error occurred while retrieving the message content");
        }
        // The consumer will continue to receive messages from the server
        // once a total of 10(prefetchCount) messages are being acknowledged.
        var result = message->basicAck();
    }
}

```
## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Required Balo version update
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [x] Added necessary documentation  
  - [x] API documentation 
  - [ ] Module documentation in Module.md files
  - [x] Ballerina By Examples

### CODE REVIEW

File: examples/rabbitmq-consumer-with-client-acknowledgement/rabbitmq_consumer_with_client_acknowledgement.bal [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    resource function onMessage(rabbitmq:Message message) {
+        // Retrieving the text content of the message.
+        var msg = message.getTextContent();
+        if (msg is string) {
+            log:printInfo("The message received: " + msg);
+        } else {
+            log:printError("Error occurred while retrieving the message content.");
+        }
+        // Positively acknowledges a single message.
+        var result = message->basicAck();
```
[the remark this answers is not in the document]
Comment (aashikam) [reply examples/rabbitmq-consumer-with-client-acknowledgement/rabbitmq_consumer_with_client_acknowledgement.bal]: Resolved

File: examples/rabbitmq-consumer-with-qos-settings/rabbitmq_consumer_with_qos_settings.bal [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    resource function onMessage(rabbitmq:Message message) {
+        var msg = message.getTextContent();
+        if (msg is string) {
+            log:printInfo("The message received: " + msg);
+        } else {
+            log:printError("Error occurred while retrieving the message content");
+        }
+        // The consumer will continue to receive messages from the server
+        // once a total of 10(prefetchCount) messages are being acknowledged.
+        var result = message->basicAck();
```
[the remark this answers is not in the document]
Comment (aashikam) [reply examples/rabbitmq-consumer-with-qos-settings/rabbitmq_consumer_with_qos_settings.bal]: Resolved

File: stdlib/messaging/rabbitmq/src/main/java/org/ballerinalang/messaging/rabbitmq/nativeimpl/channel/listener/SetQosSettings.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+import org.ballerinalang.model.values.BMap;
+import org.ballerinalang.model.values.BValue;
+import org.ballerinalang.natives.annotations.BallerinaFunction;
+import org.ballerinalang.natives.annotations.Receiver;
+
+import java.io.IOException;
+
+/**
+ *
+ */
```
[the remark this answers is not in the document]
Comment (aashikam) [reply stdlib/messaging/rabbitmq/src/main/java/org/ballerinalang/messaging/rabbitmq/nativeimpl/channel/listener/SetQosSettings.java]: Resolved

File: stdlib/messaging/rabbitmq/src/main/java/org/ballerinalang/messaging/rabbitmq/nativeimpl/channel/listener/SetQosSettings.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+            if (isValidCount && isValidSize) {
+                channel.basicQos(Math.toIntExact(((BInteger) prefetchSize).intValue()),
+                        Math.toIntExact(((BInteger) prefetchCount).intValue()),
+                        true);
+                channelObj.addNativeData(RabbitMQConstants.QOS_STATUS, true);
+            } else if (isValidCount) {
+                channel.basicQos(Math.toIntExact(((BInteger) prefetchCount).intValue()),
+                        true);
+                channelObj.addNativeData(RabbitMQConstants.QOS_STATUS, true);
+            }
```
[the remark this answers is not in the document]
Comment (aashikam) [reply stdlib/messaging/rabbitmq/src/main/java/org/ballerinalang/messaging/rabbitmq/nativeimpl/channel/listener/SetQosSettings.java]: Resolved

File: stdlib/messaging/rabbitmq/src/main/java/org/ballerinalang/messaging/rabbitmq/nativeimpl/channel/listener/Start.java [hunk 5]
```diff
@@ -85,170 +82,124 @@ public void execute(Context context) {
...
-
-        @Override
-        public void notifySuccess() {
-            // nothing to handle
-        }
-
-        @Override
-        public void notifyFailure(BError error) {
-            ErrorHandlerUtils.printError("error: " + BLangVMErrors.getPrintableStackTrace(error));
+            boolean qosStatus = channelObj.getNativeData(RabbitMQConstants.QOS_STATUS) != null;
```
[the remark this answers is not in the document]
Comment (aashikam) [reply stdlib/messaging/rabbitmq/src/main/java/org/ballerinalang/messaging/rabbitmq/nativeimpl/channel/listener/Start.java]: Resolved

File: examples/rabbitmq-consumer-with-qos-settings/rabbitmq_consumer_with_qos_settings.out [hunk 6]
```diff
@@ -0,0 +1,10 @@
+# Make sure to have the RabbitMQ broker running.
+#
+# Navigate to the directory that contains the
+# 'rabbitmq-simple-consumer.bal' file and issue the 'ballerina run' command as follows..
+
+$ ballerina run rabbitmq-simple-consumer.bal
```
Comment (daneshk) [inline examples/rabbitmq-consumer-with-qos-settings/rabbitmq_consumer_with_qos_settings.out]:
```suggestion
$ ballerina run rabbitmq_consumer_with_qos_settings.bal
```

File: stdlib/messaging/rabbitmq/src/main/java/org/ballerinalang/messaging/rabbitmq/nativeimpl/channel/listener/Start.java [hunk 7]
```diff
@@ -85,170 +82,124 @@ public void execute(Context context) {
...
+                                  boolean autoAck) {
+        ProgramFile programFile = resource.getResourceInfo().getPackageInfo().getProgramFile();
+        BMap<String, BValue> messageObj = BLangConnectorSPIUtil.createBStruct(
+                programFile, RabbitMQConstants.PACKAGE_RABBITMQ, RabbitMQConstants.MESSAGE_OBJECT);
+        messageObj.addNativeData(RabbitMQConstants.DELIVERY_TAG, deliveryTag);
+        messageObj.addNativeData(RabbitMQConstants.CHANNEL_NATIVE_OBJECT, channel);
+        messageObj.addNativeData(RabbitMQConstants.MESSAGE_CONTENT, message);
+        if (autoAck) {
+            messageObj.addNativeData(RabbitMQConstants.AUTO_ACK_STATUS, true);
         }
```
Comment (daneshk) [inline stdlib/messaging/rabbitmq/src/main/java/org/ballerinalang/messaging/rabbitmq/nativeimpl/channel/listener/Start.java]: Shall we remove the if clause and set the value like,
messageObj.addNativeData(RabbitMQConstants.AUTO_ACK_STATUS, autoAck);

File: stdlib/messaging/rabbitmq/src/main/java/org/ballerinalang/messaging/rabbitmq/nativeimpl/channel/listener/Start.java [hunk 8]
```diff
@@ -85,170 +82,124 @@ public void execute(Context context) {
...
-    /**
-     * Dispatch messages in JSON format.
-     *
-     * @param resource Ballerina resource function.
-     * @param channel  RabbitMQ channel.
-     * @param queue    Name of the queue.
-     */
-    private void dispatchJsonMessage(Resource resource, Channel channel, String queue) {
+        boolean isValidPrefetchSize = annotationValue.getRefField(RabbitMQConstants.PREFETCH_SIZE) != null;
+        boolean isValidGlobal = annotationValue.getRefField(RabbitMQConstants.PREFETCH_GLOBAL) != null;
```
Comment (daneshk) [inline stdlib/messaging/rabbitmq/src/main/java/org/ballerinalang/messaging/rabbitmq/nativeimpl/channel/listener/Start.java]: Do we need `global` parameter now?

File: examples/rabbitmq-consumer-with-qos-settings/rabbitmq_consumer_with_qos_settings.out (hunk 6 above)
Comment (aashikam) [reply examples/rabbitmq-consumer-with-qos-settings/rabbitmq_consumer_with_qos_settings.out]: Resolved

File: stdlib/messaging/rabbitmq/src/main/java/org/ballerinalang/messaging/rabbitmq/nativeimpl/channel/listener/Start.java (hunk 7 above)
Comment (aashikam) [reply stdlib/messaging/rabbitmq/src/main/java/org/ballerinalang/messaging/rabbitmq/nativeimpl/channel/listener/Start.java]: Resolved

File: examples/rabbitmq-consumer-with-client-acknowledgement/rabbitmq_consumer_with_client_acknowledgement.description [hunk 9]
```diff
@@ -0,0 +1,5 @@
+// In this example, the messages are consumed from an
+// already existing queue using the Ballerina RabbitMQ message listener.
```
Comment (praneesha) [inline examples/rabbitmq-consumer-with-client-acknowledgement/rabbitmq_consumer_with_client_acknowledgement.description]:
```suggestion
// existing queue using the Ballerina RabbitMQ message listener.
```

File: examples/rabbitmq-consumer-with-client-acknowledgement/rabbitmq_consumer_with_client_acknowledgement.description [hunk 10]
```diff
@@ -0,0 +1,5 @@
...
+// The received messages are acknowledged manually.
+// ackMode is by default rabbitmq:AUTO_ACK which will automatically acknowledge
```
Comment (praneesha) [inline examples/rabbitmq-consumer-with-client-acknowledgement/rabbitmq_consumer_with_client_acknowledgement.description]:
```suggestion
// By default, the ackMode is rabbitmq:AUTO_ACK, which will automatically acknowledge
```

File: examples/rabbitmq-consumer-with-qos-settings/rabbitmq_consumer_with_qos_settings.description [hunk 11]
```diff
@@ -0,0 +1,14 @@
+// In this example, the messages are consumed from an
+// already existing queue using the Ballerina RabbitMQ message listener,
```
Comment (praneesha) [inline examples/rabbitmq-consumer-with-qos-settings/rabbitmq_consumer_with_qos_settings.description]:
```suggestion
// existing queue using the Ballerina RabbitMQ message listener,
```

File: examples/rabbitmq-consumer-with-qos-settings/rabbitmq_consumer_with_qos_settings.description [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+// requesting specific "quality of service" settings.
+// These settings(prefetchCount and prefetchSize) impose limits on the amount
```
Comment (praneesha) [inline examples/rabbitmq-consumer-with-qos-settings/rabbitmq_consumer_with_qos_settings.description]:
```suggestion
// These settings (i.e., prefetchCount and prefetchSize) impose limits on the amount
```

File: examples/rabbitmq-consumer-with-qos-settings/rabbitmq_consumer_with_qos_settings.description [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+// of data the server will deliver to consumers before requiring acknowledgements.
+// Thus they provide a means of consumer-initiated flow control.
```
Comment (praneesha) [inline examples/rabbitmq-consumer-with-qos-settings/rabbitmq_consumer_with_qos_settings.description]:
```suggestion
// Thus, they provide a means of consumer-initiated flow control.
```

File: examples/rabbitmq-consumer-with-qos-settings/rabbitmq_consumer_with_qos_settings.description [hunk 14]
```diff
@@ -0,0 +1,14 @@
...
+// prefetchSize is the maximum amount of content (measured in octets)
+// that the server will deliver, 0 if unlimited.
```
Comment (praneesha) [inline examples/rabbitmq-consumer-with-qos-settings/rabbitmq_consumer_with_qos_settings.description]:
```suggestion
// that the server will deliver. If the amount of content is unlimited, its value will be zero.
```

File: examples/rabbitmq-consumer-with-qos-settings/rabbitmq_consumer_with_qos_settings.description [hunk 15]
```diff
@@ -0,0 +1,14 @@
...
+// prefetchCount is the maximum number of messages that the server will deliver,
```
Comment (praneesha) [inline examples/rabbitmq-consumer-with-qos-settings/rabbitmq_consumer_with_qos_settings.description]:
```suggestion
// prefetchCount is the maximum number of messages that the server will deliver.
```

File: examples/rabbitmq-consumer-with-qos-settings/rabbitmq_consumer_with_qos_settings.description [hunk 16]
```diff
@@ -0,0 +1,14 @@
...
+// 0 if unlimited.
```
Comment (praneesha) [inline examples/rabbitmq-consumer-with-qos-settings/rabbitmq_consumer_with_qos_settings.description]:
```suggestion
// If the amount of content is unlimited, its value will be zero.
```

File: examples/rabbitmq-consumer-with-qos-settings/rabbitmq_consumer_with_qos_settings.description [hunk 17]
```diff
@@ -0,0 +1,14 @@
...
+// If the settings are specified at the listener initialization, they will be
+// applied for the entire connection, as a global setting.
```
Comment (praneesha) [inline examples/rabbitmq-consumer-with-qos-settings/rabbitmq_consumer_with_qos_settings.description]:
```suggestion
// applied for the entire connection as a global setting.
```

**Pull request merged**